Add tutorial for how to improve public instance + rework current pages (#464)
iv-org/documentation: docs/improve-public-instance.md

# Improve your public instance performance and stability

This tutorial has been written by [unixfox](https://github.com/unixfox), owner of [yewtu.be](https://yewtu.be/). He is better suited when looking for help about this tutorial.

> ## Synopsis

Author: I'm keeping Synopsis it's valid :D

> ## Synopsis

Reviewer: A synopsis is used for movies and stories, not really for technical documents.... so it's not great.

   **Don't restart Invidious yet!**
3. Create a file called `nginx.conf` and put this content:
   ```
   user www-data;
   events {
        worker_connections 1000;
   }
   http {
        server {
            listen 3000;
            location / {
                proxy_pass http://invidious:3000;
            }
        }
   }
   ```
4. Add a new service in your `docker-compose.yml` file:
   ```
   nginx:
      image: nginx:latest
      restart: unless-stopped
      volumes:
        - ./nginx.conf:/etc/nginx/nginx.conf:ro
      depends_on:
        - invidious
      ports:
        - "127.0.0.1:3000:3000"
   ```

5. Update your CRON for restarting Invidious.
   Instead of restarting a single Docker container you will now need to restart 6 containers (adjust if you added or removed number of containers).
   Replace with these CRON lines:
   ```
   0 */1 * * * docker restart invidious-invidious-1
   1 */1 * * * docker restart invidious-invidious-2
   2 */1 * * * docker restart invidious-invidious-3
   3 */1 * * * docker restart invidious-invidious-4
   4 */1 * * * docker restart invidious-invidious-5
   5 */1 * * * docker restart invidious-invidious-6
   ```

>    Each CRON line has a different schedule for avoiding disrupting your entire Invidious instance due a restart.

Reviewer: ```suggestion
   Each CRON line has a different schedule to avoid disrupting your entire Invidious instance due a restart.
```

>    Each CRON line has a different schedule for avoiding disrupting your entire Invidious instance due a restart.

Reviewer 2: ```suggestion
   Each CRON line has a different schedule to avoid disrupting your entire Invidious instance due a restart.
```

6. Apply the new configuration:
   ```
   docker compose down
   chmod +x ./start.sh
   ./start.sh
   ```

> ### 2) Accelerate the video playback with http3-ytproxy

Reviewer: ```suggestion
### 2) Speed up videos playback with http3-ytproxy
```

> Kavin from Piped has developped a tool that is much faster at proxying the traffic for video playback and image loading than Invidious: https://github.com/TeamPiped/http3-ytproxy

Reviewer 2: ```suggestion
Kavin from Piped has developed a tool that is much faster at proxying the traffic for video playback and image loading than Invidious: https://github.com/TeamPiped/http3-ytproxy
```

> Kavin from Piped has developped a tool that is much faster at proxying the traffic for video playback and image loading than Invidious: https://github.com/TeamPiped/http3-ytproxy

Reviewer: ```suggestion
Kavin from the Piped team has developed a tool that is much faster at proxying the traffic for video playback and image loading than Invidious: https://github.com/TeamPiped/http3-ytproxy
```

NGINX configuration will be used for this tutorial, and it's highly recommended to setup this configuration on your main reverse proxy.

> But if you do not have NGINX as your main reverse proxy you can either try to adapt the rules to your reverse proxy. Or you can also use the separate NGINX container from the [first section](#1-multiple-invidious-processes).

Reviewer: The previous setup implies that NGINX is used, it's not great to imply something else here.

Author: no the previous setup do not require NGINX, NGINX is used as a middleman

Reviewer: Oh... ok. That's a bit confusing then, maybe add a warning that we're not talking about the middleman setup previously.

Author: that's why I wrote "main reverse proxy". I made it so it's in bold now.

Reviewer: Yeah, sure I did see it, but it's a bit confusing. I guess we'll see if people get confused when they set it up.

> 1. Create a new directory for this new service and set it all permissions:

Reviewer: ```suggestion
1. Create a new directory for this new service and give it all permissions:
```

   ```
   mkdir /opt/http3-ytproxy

>    chmod 777 -R /opt/http3-ytproxy

Reviewer: 777 is bad practice, would be nice to know who's supposed to be the owner and just chown the folder

Author: It's ok because the actual interesting file is created with the correct permissions. And it's not a mission-critical socket, it's just a proxy.
But ok I'll instead chown the directory.

   ```

> 2. Find the username of NGINX process, you can find it at the top of the file `/etc/nginx/nginx.conf`.  

Reviewer: Same as before, the username was set to www-data in the config we give previously

   ```
   http3-ytproxy:
      image: 1337kavin/ytproxy:latest
      restart: unless-stopped
      user: 33:33
      environment:
        - DISABLE_WEBP=1
      volumes:
        - /opt/http3-ytproxy:/app/socket
   ```
   Replace `33:33` with uid:gid if you have something different.
4. Add these lines for the "server" section of Invidious in your NGINX configuration, just after the block `location /`:
   ```
   location ~ (^/videoplayback|^/vi/|^/ggpht/|^/sb/) {
        proxy_buffering on;
        proxy_buffers 1024 16k;
        proxy_set_header X-Forwarded-For "";
        proxy_set_header CF-Connecting-IP "";
        proxy_hide_header "alt-svc";
        sendfile on;
        sendfile_max_chunk 512k;
        tcp_nopush on;
        aio threads=default;
        aio_write on;
        directio 16m;
        proxy_hide_header Cache-Control;
        proxy_hide_header etag;
        proxy_http_version 1.1;
        proxy_set_header Connection keep-alive;
        proxy_max_temp_file_size 32m;
        access_log off;
        proxy_pass http://unix:/opt/http3-ytproxy/http-proxy.sock;
        add_header Cache-Control private always;
   }
   ```

>    If using the NGINX from the [first section](#1-multiple-invidious-processes), you will need to add this new volume:

Reviewer: ```suggestion
   If you're using the NGINX from the [first section](#1-multiple-invidious-processes), you will need to add this new volume:
```
